encoder/v1/encoder.go

```diff
@@ -0,0 +1,154 @@
+package v1
+
+import (
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+
+	encoders "github.com/smartcontractkit/ocr2keepers/encoder"
+	"github.com/smartcontractkit/ocr2keepers/pkg/chain"
+	"github.com/smartcontractkit/ocr2keepers/pkg/types"
+)
+
+const (
+	separator = "|"
+)
+
+var (
+	errUpkeepKeyNotParsable = fmt.Errorf("upkeep key not parsable")
+)
+
+type wrappedPerform struct {
+	CheckBlockNumber uint32   `abi:"checkBlockNumber" json:"checkBlockNumber"`
+	CheckBlockhash   [32]byte `abi:"checkBlockhash" json:"checkBlockhash"`
+	PerformData      []byte   `abi:"performData" json:"performData"`
+}
+
+type reportPacker interface {
+	Pack(args ...interface{}) ([]byte, error)
+	UnpackIntoMap(v map[string]interface{}, data []byte) error
+}
+
+type encoder struct {
+	validator
+	upkeepProvider
+	eligibilityProvider
+	reportKeys []string
+	packer     reportPacker
+}
+
+func NewEncoder() *encoder {
+	arguments := abi.Arguments{
+		{Name: "fastGasWei", Type: chain.Uint256},
+		{Name: "linkNative", Type: chain.Uint256},
+		{Name: "upkeepIds", Type: chain.Uint256Arr},
+		{Name: "wrappedPerformDatas", Type: chain.PerformDataArr},
+	}
+
+	reportKeys := make([]string, len(arguments))
+	for i, arg := range arguments {
+		reportKeys[i] = arg.Name
+	}
+
+	return &encoder{
+		reportKeys: reportKeys,
+		packer:     arguments,
+	}
+}
+
+// EncodeReport encodes a report from a list of upkeep results
+func (e *encoder) EncodeReport(toReport []types.UpkeepResult, _ ...encoders.Config) ([]byte, error) {
+	if len(toReport) == 0 {
+		return nil, nil
+	}
+
+	var baseValuesIdx int
+	for i, rpt := range toReport {
+		if rpt.CheckBlockNumber > uint32(baseValuesIdx) {
+			baseValuesIdx = i
+		}
```

Comment:
I don't fully follow this logic, why do we compare checkBlockNumber with an index value?

Reply:
It seems we want to use one value for fastgas and link native. I would say we should use the one with highest checkBlockNumber?

Reply:
This was raised in the original PR for this change: https://github.com/smartcontractkit/ocr2keepers/pull/95#discussion_r1138884835 But it still needs to be addressed